Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1241 @@
+        }
+        else
+        {
+            m_value = AZStd::allocate_shared<const AZStd::string>(AZStdAlloc<ValueAllocator>(), value);
+        }
+    }
+
+    AZStd::any& Value::GetOpaqueValue() const
+    {
+        return *AZStd::get<AZStd::any*>(m_value);
+    }
+
+    void Value::SetOpaqueValue(AZStd::any& value)
+    {
+        m_value = &value;
+    }
+
+    void Value::SetNull()
+    {
+        m_value = AZStd::monostate();
+    }
+
+    Visitor::Result Value::Accept(Visitor& visitor, bool copyStrings) const
+    {
+        Visitor::Result result = AZ::Success();
+
+        AZStd::visit(
+            [&](auto&& arg)
+            {
+                using Alternative = AZStd::decay_t<decltype(arg)>;
+
+                if constexpr (AZStd::is_same_v<Alternative, AZStd::monostate>)
+                {
+                    result = visitor.Null();
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, int64_t>)
+                {
+                    result = visitor.Int64(arg);
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, uint64_t>)
+                {
+                    result = visitor.Uint64(arg);
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, double>)
+                {
+                    result = visitor.Double(arg);
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, bool>)
+                {
+                    result = visitor.Bool(arg);
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, AZStd::string_view>)
+                {
+                    result = visitor.String(arg, copyStrings ? Lifetime::Temporary : Lifetime::Persistent);
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, AZStd::shared_ptr<const AZStd::string>>)
+                {
+                    result = visitor.RefCountedString(arg, copyStrings ? Lifetime::Temporary : Lifetime::Persistent);
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, ObjectPtr>)
+                {
+                    result = visitor.StartObject();
+                    if (result.IsSuccess())
+                    {
+                        const Object::ContainerType& object = GetObjectInternal();
+                        for (const Object::EntryType& entry : object)
+                        {
+                            result = visitor.Key(entry.first);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                            result = entry.second.Accept(visitor, copyStrings);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                        }
+                        result = visitor.EndObject(object.size());
```

Comment:
If for some reason tehe visiting of the object member fails, this code does not call `EndObject`,

This shouldn't return, but instead break the for loop, to make sure that EndObject is invoked, if a successful StartObject has been invoked

Reply:
This is a general issue across the implementation of the generic DOM visitor implementations because I didn't think about recoverable/partial failures. I'd like to leave this as-is for now and take a follow-up task to fix this everywhere (using something similar to our JSON serializer's `MergeResult` pattern) if possible